casper/src/test/resources/PoSTest.rho

```diff
@@ -3,32 +3,290 @@
+              prepareUser!("6666666666666666666666666666666666666666666666666666666666666666", 10000, *setupCh) |
+              for (_, _, _ <- setupCh) {
+                @PoS!("getBonds", *bondsCh) |
+                for (@initialBonds <- bondsCh) {
+                  @PoS!("bond", -1, *retCh) |
+                  for ( @(result, _) <- retCh) {
+                    @PoS!("getBonds", *bondsCh) |
+                    rhoSpec!("assertMany",
+                      [
+                        ((result, "==", false), "the bond should fail"),
+                        ((initialBonds, "== <-", *bondsCh), "the bonds map remains unchanged")
+                      ],
+                      *ackCh
+                    )
+                  }
+                }
+              }
+            }
+          } |
 
-        contract setup(returnCh) = {
-          returnCh!(Nil)
-        } |
+          contract test_bonding_fails_if_bond_too_big(rhoSpec, _, ackCh) = {
+            new setupCh, retCh, bondsCh in {
+              prepareUser!("7777777777777777777777777777777777777777777777777777777777777777", 10000, *setupCh) |
+              for (_, _, _ <- setupCh) {
+                @PoS!("getBonds", *bondsCh) |
+                for (@initialBonds <- bondsCh) {
+                  @PoS!("bond", 100000, *retCh) |
+                  for ( @(result, _) <- retCh) {
+                    @PoS!("getBonds", *bondsCh) |
+                    rhoSpec!("assertMany",
+                      [
+                        ((result, "==", false), "the bond should fail"),
+                        ((initialBonds, "== <-", *bondsCh), "the bonds map remains unchanged")
+                      ],
+                      *ackCh
+                    )
+                  }
+                }
+              }
+            }
+          } |
 
-        contract test(rhoSpec, _, ackCh) = {
-          rhoSpec!("assert", (Nil, "!=", PoS), "there is PoS", *ackCh)
+          contract test_bonding_fails_if_already_bonded(rhoSpec, _, ackCh) = {
+            new setupCh, initialBondsCh, bond1Ch, bond2Ch, bondsCh in {
+              prepareUser!("8888888888888888888888888888888888888888888888888888888888888888", 10000, *setupCh) |
+              for (@user1PubKey, _, _ <- setupCh) {
+                @PoS!("bond", 150, *bond1Ch) |
+                for ( @(result1, _) <- bond1Ch) {
+                  @PoS!("getBonds", *initialBondsCh) |
+                  for ( @initialBonds <- initialBondsCh) {
+                    @PoS!("bond", 150, *bond2Ch) |
+                    for ( @(result2, _) <- bond2Ch) {
+                      @PoS!("getBonds", *bondsCh) |
+                      rhoSpec!("assertMany",
+                        [
+                          ((result1, "==", true), "the first bond should succeed"),
+                          ((result2, "==", false), "the second bond should fail"),
+                          ((initialBonds, "== <-", *bondsCh), "the bonds map remains unchanged")
```

Comment:
The 'bonding fails' scenario could be extracted into a method and reused 4 times. In this case, we'd do the initial successful bonding, an assertion about its success, and then we'd delegate to `testBondFailure(150, "Already bonded!", ackCh)`

Reply:
It can be reused 3 times, indeed. The `test_bonding_fails_if_already_bonded` is different